### PR TITLE
Add apps

### PR DESCRIPTION
**Pull request by mitechie**
*TO QA THIS you'll need to reload the whole db*

This adds support for tracking hits of applications to the db. We add an `application_hits` db table with the charmid, charm name, owner, series, along with uuid/day. 

QA:

You should rebuild the python env so that you get the new dependency. 

    make clean-all
    make 
    python longrunning/longrunning.py initdb
    ....
    
    sqlite3 models.db
    .tables
    SELECT * FROM application_hits LIMIT 50;

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -31,6 +31,7 @@ sysdeps:
 	$(PIP) install click
 	$(PIP) install python-swiftclient
 	$(PIP) install python-keystoneclient
+	$(PIP) install jujubundlelib
```
Comment (mitechie) [inline Makefile:34]: Added for the Reference class to parse the charm id for series/owner metadata.

File: long-running/longrunning.py [hunk 2]
```diff
@@ -114,7 +134,8 @@ def find_uuid(l):
     m = re.search(uuid_re, l)
     if m:
         uuid = m.group(0)
-        return uuid.split(b'=')[1]
+        var = uuid.split(b'=')[1].decode('utf-8')
```
Comment (mitechie) [inline long-running/longrunning.py]: queries fail if bytes go into the db so all values need to be decoded.

File: long-running/longrunning.py [hunk 3]
```diff
@@ -261,7 +332,7 @@ def output_latest_day_versions(conn):
     print("\n\n{} saw:".format(day))
     print("Count\tVersion")
     for row in versions:
-        print("{0}\t{1}".format(row[0], row[1].decode('utf-8')))
```
Comment (mitechie) [inline long-running/longrunning.py:264]: the values are decoded in the original parsing code vs the output code now.

File: long-running/longrunning.py [hunk 4]
```diff
@@ -270,7 +341,7 @@ def output_latest_day_clouds(conn):
     print("\n\n{} saw:".format(day))
     print("Count\tCloud")
     for row in clouds:
-        print("{0}\t{1}".format(row[0], row[1].decode('utf-8')))
```
Comment (mitechie) [inline long-running/longrunning.py:273]: the values are decoded in the original parsing code vs the output code now.

**Comment by bac** [comment]
clean_venv target should just remove .venv. It didn't get updated from the original version.

File: Makefile (hunk 1 above)
Comment (bac) [reply Makefile:34]: Consider a requirements.pip file?

File: long-running/longrunning.py [hunk 5]
```diff
@@ -7,6 +7,10 @@
 import re
 import sqlite3
 
+from collections import namedtuple
+from urllib.parse import unquote
+from jujubundlelib.references import Reference
```
Comment (bac) [inline long-running/longrunning.py]: Maybe do standard grouping and alphabetize? meh.

File: long-running/longrunning.py (hunk 2 above)
Comment (bac) [reply long-running/longrunning.py]: That'd make a great code comment, you know, for posterity. :)

File: long-running/longrunning.py [hunk 6]
```diff
@@ -136,7 +157,42 @@ def find_metadata(l):
...
+    if app_raw:
+        _, charmid = app_raw.group().split(b'=')
+        charmid = unquote(charmid.decode("utf-8"))
+
+        # Use the jujubundlelib to parse the charm url for the series
+        try:
+            ref = Reference.from_string(charmid)
+        except ValueError:
+            # skip things if there's an error parsing the charm url
+            return None
```
Comment (bac) [inline long-running/longrunning.py]: Log the offending charmid?

**Comment by bac** [comment]
In `longrunning.py` you have

```
    for g in logs:
        print("Found logs {0}".format(len(g)))
        for path in g:
            print("Processing: {0}".format(path))
```

The glob definition of logs returns paths in unsorted order, so they get processed all willy nilly. Would make a nicer user experience if the paths were sorted so there would be a sense of progress.

**Review by bac — APPROVED** [review]

File: long-running/longrunning.py (hunk 5 above)
Comment (mitechie) [reply long-running/longrunning.py]: oops was meant to be updated.

File: long-running/longrunning.py (hunk 6 above)
Comment (mitechie) [reply long-running/longrunning.py]: +1

**Comment by mitechie** [comment]
Updated per feedback. TY for the review and long long QA!

**Pull request merged**